tests/php/Forms/FormFieldTest.php

```diff
@@ -455,6 +475,106 @@ public function testGetSchemaStateWithFormValidation()
+        foreach ($formFieldClasses as $formFieldClass) {
+            $reflection = new ReflectionClass($formFieldClass);
+            // Skip abstract classes, like MultiSelectField, and fields that only exist for unit tests
+            if ($reflection->isAbstract() || is_a($formFieldClass, TestOnly::class, true)) {
+                continue;
+            }
+
+            switch ($formFieldClass) {
+                case NullableField::class:
+                case CompositeField::class:
+                case FieldGroup::class:
+                case PopoverField::class:
+                    $args = [TextField::create('Test2')];
+                    break;
+                case SelectionGroup_Item::class:
+                    $args = ['Test', [TextField::create('Test2')]];
+                    break;
+                case ToggleCompositeField::class:
+                    $args = ['Test', 'Test', TextField::create('Test2')];
+                    break;
+                case PrintableTransformation_TabSet::class:
+                    $args = [Tab::create('TestTab', 'Testtab', TextField::create('Test2'))];
+                    break;
+                case TreeDropdownField::class:
+                case TreeDropdownField_Readonly::class:
+                    $args = ['Test', 'Test', Group::class];
+                    break;
+                case PermissionCheckboxSetField::class:
+                case PermissionCheckboxSetField_Readonly::class:
+                    $args = ['Test', 'Test', Permission::class, 'Test'];
+                    break;
+                case SelectionGroup::class:
+                    $args = ['Test', []];
+                    break;
+                case GridField_FormAction::class:
+                    $args = [GridField::create('GF'), 'Test', 'Test label', 'Test action name', []];
+                    break;
+                case GridState::class:
+                    $args = [GridField::create('GF')];
+                    break;
+                default:
+                    $args = ['Test', 'Test label'];
+            }
+
+            // Assert that extendValidationResult is called once each time ->validate() is called
+            $mock = $this->getMockBuilder($formFieldClass)
+                ->setConstructorArgs($args)
+                ->onlyMethods(['extendValidationResult'])
+                ->getMock();
+            $mock->expects($invocationRule = $this->once())
+                ->method('extendValidationResult')
+                ->will($this->returnValue(true));
+
+            $isValid = $mock->validate(new RequiredFields());
+            $this->assertTrue($isValid, "$formFieldClass should be valid");
+
+            // This block is not essential and only exists to make test debugging easier - without this,
+            // the error message on failure is generic and doesn't include the class name that failed
+            try {
+                $invocationRule->verify();
```

Comment:
This API isn’t covered by phpunit’s BC promise, but it hasn’t changed from 8.x to 10.x: https://github.com/sebastianbergmann/phpunit/blob/10.0.9/src/Framework/MockObject/Rule/InvokedCount.php. I’ve included the comment for info so if it causes failures in future, this block can just be removed and the test will still run fine, just with less debugging info on error.